File: tests/test-unix-socket.c
/*
 * Copyright (c) 2010, 2012, 2014 Nicira, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at:
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <config.h>
#undef NDEBUG
#include "socket-util.h"
#include <errno.h>
#include <signal.h>
#include <unistd.h>
#include "ovstest.h"
#include "util.h"

static void
test_unix_socket_main(int argc, char *argv[])
{
    const char *sockname1;
    const char *sockname2;
    int sock1, sock2;

    set_program_name(argv[0]);

    if (argc != 2 && argc != 3) {
        ovs_fatal(0, "usage: %s SOCKETNAME1 [SOCKETNAME2]", argv[0]);
    }
    sockname1 = argv[1];
    sockname2 = argc > 2 ? argv[2] : sockname1;

    signal(SIGALRM, SIG_DFL);
    alarm(5);

    /* Create a listening socket under name 'sockname1'. */
    sock1 = make_unix_socket(SOCK_STREAM, false, sockname1, NULL);
    if (sock1 < 0) {
        ovs_fatal(-sock1, "%s: bind failed", sockname1);
    }
    if (listen(sock1, 1)) {
        ovs_fatal(errno, "%s: listen failed", sockname1);
    }

    /* Connect to 'sockname2' (which should be the same file, perhaps under a
     * different name). */
    sock2 = make_unix_socket(SOCK_STREAM, false, NULL, sockname2);
    if (sock2 < 0) {
        ovs_fatal(-sock2, "%s: connect failed", sockname2);
    }

    close(sock1);
    close(sock2);
}

OVSTEST_REGISTER("test-unix-socket", test_unix_socket_main);

File: tests/test-unix-socket.py
#
# Copyright (c) 2010, 2012, 2013 Nicira, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at:
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import os
import signal
import socket
import sys

import ovs.socket_util
from ovs.fatal_signal import signal_alarm


def main(argv):
    if len(argv) not in (2, 3):
        sys.stderr.write("usage: %s SOCKETNAME1 [SOCKETNAME2]", argv[0])
        sys.exit(1)

    sockname1 = argv[1]
    if len(argv) > 2:
        sockname2 = argv[2]
    else:
        sockname2 = sockname1

    signal.signal(signal.SIGALRM, signal.SIG_DFL)
    signal_alarm(5)

    # Create a listening socket under name 'sockname1'.
    error, sock1 = ovs.socket_util.make_unix_socket(socket.SOCK_STREAM, False,
                                                    sockname1, None)
    if error:
        sys.stderr.write("%s: bind failed (%s)" % (sockname1,
                                                   os.strerror(error)))
        sys.exit(1)
    sock1.listen(1)

    # Connect to 'sockname2' (which should be the same file, perhaps under a
    # different name).
    error, sock2 = ovs.socket_util.make_unix_socket(socket.SOCK_STREAM, False,
                                                    None, sockname2)
    if error:
        sys.stderr.write("%s: connect failed (%s)" % (sockname2,
                                                      os.strerror(error)))
        sys.exit(1)


if __name__ == '__main__':
    main(sys.argv)
